odoo/fields.py

```diff
@@ -3070,6 +3075,676 @@ def _record_ids_per_res_model(self, records):
+                'value': 'red',
+            }, {
+                'name': 'aa34746a6851ee4ea1f8d95746e45788',
+                'string': 'Partner',
+                'type': 'many2one',
+                'comodel': 'test_new_api.partner',
+                'value': [1337, 'Bob'],
+            }]
+
+            Output dict:
+            {
+                '3adf37f3258cfe40f0907d9cbdd7d091': 'red',
+                'aa34746a6851ee4ea1f8d95746e45788': 1337,
+            }
+
+        :param values_list: List of properties definition and value
+        :return: Generate a dict {name: value} from this definitions / values list
+        """
+        if not is_list_of(values_list, dict):
+            raise ValueError(f'Wrong properties value {values_list!r}')
+
+        dict_value = {}
+        for property_definition in values_list:
+            property_value = property_definition.get('value')
+            property_type = property_definition.get('type')
+            property_model = property_definition.get('comodel')
+
+            if property_type in ('many2one', 'many2many') and property_model and property_value:
+                # check that value are correct before storing them in database
+                if property_type == 'many2many' and property_value and not is_list_of(property_value, int):
+                    raise ValueError(f"Wrong many2many value {property_value!r}")
+
+                elif property_type == 'many2one' and not isinstance(property_value, int):
+                    raise ValueError(f"Wrong many2one value {property_value!r}")
+
+            dict_value[property_definition['name']] = property_value
+
+        return dict_value
+
+    @classmethod
+    def _dict_to_list(cls, values_dict, properties_definition):
+        """Convert a dict of {property: value} into a list of property definition with values.
+
+        :param values_dict: JSON value coming from the child table
+        :param properties_definition: Properties definition coming from the parent table
+        :return: Merge both value into a list of properties with value
+            Ignore every values in the child that is not defined on the parent.
+        """
+        if not is_list_of(properties_definition, dict):
+            raise ValueError(f'Wrong properties value {properties_definition!r}')
+
+        values_list = copy.deepcopy(properties_definition)
+        for property_definition in values_list:
+            property_definition['value'] = values_dict.get(property_definition['name'])
+        return values_list
+
+
+class PropertiesDefinition(Field):
+    """ Field used to define the properties definition (see :class:`~odoo.fields.Properties`
+    field). This field is used on a "parent" to define the structure of
+    expected properties on "children". It is used to verify the properties
+    definition. """
+    type = 'properties_definition'
+    column_type = ('jsonb', 'jsonb')
+    copy = False
+    readonly = False
+    prefetch = False
```

Comment:
The issue is that prefetching goes both ways.  If you access another prefetchable field, this one is going to be fetched in the same query.  On the other hand, if you access this field (which is what the properties field does on other records), it is going to fetch all other prefetchable fields as well.  One may mitigate this by making the field prefetchable, but add `prefetch_fields=False` in the context when accessing the definition record...

My intuition goes along these lines:
* business code using the definition record (or "parent" record) will normally not need the definition field, as it has no direct business value;
* business code accessing the properties field on a record will possibly only access the definition field on the definition record.

Having `prefetch = True` by default makes sense if accessing several fields a "child" record implies accessing several fields on the "parent" record, which I think is not the case in general.  I have the feeling that properties fields and their definition fields should be considered non-critical for performance, since we don't expect actual business code to use them.

All that being said, the _cost_ of making those fields prefetchable looks reasonable when compared to binary fields, for instance.

Reply:
After discussion: seems it can be turned on afterwards for performance purpose, but not really critical currently. As this makes runbot red (queries change), I keep it as False and we will check afterwards (added a note in the post-merge todolist).